### ISSUE TITLE
Support Partial/Filtered Unique Index for PostgreSQL and SQL Server

### ISSUE THREAD

**Issue by bkrausz** [prisma/prisma#3076]
## Problem
tl;dr - Support for [Postgres Partial Indexes](https://www.postgresql.org/docs/8.0/indexes-partial.html) and related functionality

Consider, for ex, a versioned CMS schema 
```prisma
model Page {
  id                Int       @id @default(autoincrement)
  staticId          String
  isLatest          Boolean
  previousVersionId Int?
  previousVersion   Workflow? @relation("pagePreviousVersion", fields: [previousVersionId], references: [id])
  nextVersion       Workflow  @relation("pagePreviousVersion")
}
```

New entries for the same page (i.e. identical `staticId`) are created by adding a new entry with `isLatest=true` and setting `isLatest=false` on the old one. We want to enforce that only one version is ever latest, so we have a partial index on `staticId` where `isLatest=true`.

It looks like Prisma is mostly peaceful with this except for a few things:
1. Introspect will try to create these as full unique indexes (i.e. `@@unique`), which is incorrect
2. `post.findOne` won't allow `{where: {staticId: '...', isLatest: true}}` because there's no way to tell it that this guarantees uniqueness.

## Suggested solution
Partial index support ala `@@unique([staticId], where={isLatest: true})` would be great.

## Alternatives
I can probably work around this with `findMany(...)[0]`, but flagging as a feature request.

## Additional context
There's another challenge in supporting this data model: AFAICT there is no way to update an object at the same time as connecting it. It seems like right now I can run:
```typescript
const page = prisma.page.create({
  data: {
    staticId: prevVersion.staticId,
    isLatest: false,
    previousVersion: {
      connect: {id: prevVersion.id}
    }
  }
})
prisma.page.update({
  where: { id: page.id },
  data: {
    isLatest: true,
    previousVersion: {
      update: { isLatest: false }
    }
  }
})
```

Whereas I'd ideally like to run:
```typescript
const page = prisma.page.create({
  data: {
    staticId: prevVersion.staticId,
    isLatest: true,
    previousVersion: {
      connect: {id: prevVersion.id},
      update: {isLatest: false}
    }
  }
})
```
Happy to file as a separate feature request as it seems mostly unrelated to partial indexes.

**Comment by bkrausz** [comment]
Slight update: my example `prisma.page.update` call doesn't work because of the ordering of the execution within the transaction (there's no way to make the `isLatest: false` happen first). This does, however, work with the transactionApi as you can specify ordering.

**Comment by janpio** [comment]
Can you share the SQL of the `Page` table above? Will make it a lot easier to understand.

**Comment by bkrausz** [comment]
How we're using this (we're currently moving over from Python/SQLAlchemy to Prisma):

```sql
CREATE TABLE pages (
    id SERIAL PRIMARY KEY,
    staticId character varying NOT NULL,
    isLatest boolean NOT NULL,
    previousVersionId integer REFERENCES workflows(id),
    # ...content/title/etc fields
);

CREATE UNIQUE INDEX pages_pkey ON pages(id int4_ops);
CREATE UNIQUE INDEX "idx_pages_unique_staticId" ON pages(staticId text_ops) WHERE isLatest = true;
```

## Reading
The frontend gets the page for display via
```sql
SELECT * FROM pages WHERE staticId = ${staticId} AND isLatest = true;
```

## Updating

Given a newly created page with a single revision:
| id | staticId | isLatest | previousVersionId |
|----|----------|----------|-------------------|
| 1  | page_abc | true    | NULL              |

One would create a new version of page `staticId` from current latest version `currentPageId`:

```sql
BEGIN;
UPDATE pages SET isLatest = false WHERE id = ${currentPageId};
INSERT INTO pages ("staticId", "isLatest", "previousVersionId") VALUES (${staticId}, true, ${currentPageId});
COMMIT;
```

Resulting in:
| id | staticId | isLatest | previousVersionId |
|----|----------|----------|-------------------|
| 2  | page_abc | true     | 1                 |
| 1  | page_abc | false    | NULL              |

**Comment by janpio** [comment]
Over at https://github.com/prisma/prisma/issues/2829#issuecomment-677834599 you mentioned:

> Sounds like you're on the tail end of dev work here, but any chance of getting eyes on #3076? tl;dr - reintrospection is incorrect in a particularly bad way when using partial unique indexes, having Prisma ignore them rather than incorrectly assume they're globally unique would be a large improvement in usability for us.

Can you explain the relation to (Re)Introspection here and what the current problem is? I did assume we were ignoring this completely for now.

**Comment by bkrausz** [comment]
Right now when introspection sees a partial unique index, it treats it as a unique index. Using the example from #3076, introspection would add an `@unique` to `staticId`, which must then be removed.

#3076 as a while is a larger yak, but having introspection completely ignore PUIs (which is what I currently do when manually building our schema) would be preferable to the current behavior, as the current behavior is problematically incorrect.

**Comment by janpio** [comment]
Understood. Does https://github.com/prisma/prisma/issues/3373 describe the problem properly?

**Comment by bkrausz** [comment]
Yes it does, thanks!

**Comment by janpio** [comment]
> Happy to file as a separate feature request as it seems mostly unrelated to partial indexes.

Just saw the second part of the original issue: Yes please 👍

**Comment by Sytten** [comment]
@albertoperdomo This should be revisited now

**Comment by kakserpom** [comment]
Any news?

**Comment by janpio** [comment]
On SQL Server these are called "Filtered Indexes" ([via](https://use-the-index-luke.com/sql/where-clause/partial-and-filtered-indexes)): https://docs.microsoft.com/en-us/sql/relational-databases/indexes/create-filtered-indexes?view=sql-server-ver15

**Comment by janpio** [comment]
Partial indexes are not part of the preview feature, so I can not really say if _this_ feature request here will come "soon". But we are certainly working through the index related gaps we have and closing them one by one.

**Comment by Jolg42** [comment]
Note from Tyler: This is needed as a performance optimization for large databases.

> something I used frequently in the past and would love support for are indexes with predicates. 
Example:
`CREATE INDEX IDX_User_Email_ActiveOnly ON User (email) WHERE isActive = 1`
The index is only utilized for queries with the isActive = 1 predicate, but the index size is significantly smaller than the alternative:
`CREATE INDEX IDX_User_IsActive_Email ON User (isActive, email)`
It's needed when tables are very large and an index on the whole thing is unnecessary.

**Comment by Jolg42** [comment]
Note: In a [related issue](https://github.com/prisma/prisma/issues/9885), an alternative was mentioned (of course full support for these indexes would be better) 
https://github.com/prisma/prisma/issues/16658

**Comment by JoniBR** [comment]
Any news?

**Comment by janpio** [comment]
Yes, I posted a comment in a duplicate issue asking for SQL examples so we can design this feature better: https://github.com/prisma/prisma/issues/6974#issuecomment-1483792245 Everyone who commented here or is subscribed to this issue, please go over there and comment. Thanks!

Generally, please leave comments at https://github.com/prisma/prisma/issues/6974 going forward. I am not going to close this one yet as it has a lot of upvotes, but generally #6974 is the better issue now to continue the discussion. 🙏
